client/me/purchases/cancel-purchase/button.jsx

```diff
@@ -34,7 +34,7 @@ import { getDowngradePlanFromPurchase } from 'state/purchases/selectors';
 class CancelPurchaseButton extends Component {
 	static propTypes = {
 		purchase: PropTypes.object.isRequired,
-		selectedSite: PropTypes.object.isRequired,
+		selectedSite: PropTypes.object,
```

Comment:
If you observed a `null` selectedSite here, I'd suggest auditing the entire file. For example, this is potentially problematic:

https://github.com/Automattic/wp-calypso/blob/0753a38fc4c61df18e8450abe67b84ad87c7511d/client/me/purchases/cancel-purchase/button.jsx#L79

```js
const { slug } = null;
// Throws:
// TypeError: Cannot destructure property `slug` of 'undefined' or 'null'.
```

Reply:
Do you think it would be better to keep `selectedSite` as a required prop and allow the `propType` warnings to persist? The idea being that in the rare case where a user with no sites and disconnected Jetpack purchases wants to cancel their purchase, we're ok with a warning in the console or otherwise defying this propType.

I'm operating under the assumption that warnings are something I want to avoid introducing but in this case I'm wondering what my alternatives are. I've had to remove the `siteSelection` middleware from the cancel purchase route because customers in this scenario won't have any sites to select from as it was preventing the route from being hit by displaying a "no sites" message. And so at this point in the code, I don't have a site to provide this component with.

As you mention, these warnings are weak, and I was able to cancel the purchase just fine despite getting the warning in my console. I'm wondering if that's the right thing to do. Otherwise, my only other idea would be to pass on the fact that we're dealing with a purchase without a site by passing a custom (hardcoded) object that meets the `propType` requirement, but I'm not sure I like that approach.

Reply:
> Do you think it would be better to keep `selectedSite` as a required prop and allow the `propType` warnings to persist? The idea being that in the rare case where a user with no sites and disconnected Jetpack purchases wants to cancel their purchase, we're ok with a warning in the console or otherwise defying this propType.

`propTypes` describe what we expect to find in the component. We should not see any `propTypes` warnings, they indicate a component is operating under abnormal conditions, or that our assumptions about the runtime behavior of the application are inaccurate. These are both problems and should concern us. The `propTypes` as currently written seem accurate.

From what I've understood, `selectedSite` may be `null`. You've observed `propTypes` warnings indicating that. At this point, we should have strong confidence that this component will behave correctly under all of the normal conditions:
- `this.props.selectedSite = null`
- `this.props.selectedSite = { /* …siteObject */ }`

At a glance, and I have not looked closely at the component, I would expect this line to error when `thes.props.selectedSite` is `null`:

https://github.com/Automattic/wp-calypso/blob/0753a38fc4c61df18e8450abe67b84ad87c7511d/client/me/purchases/cancel-purchase/button.jsx#L79

If that line is indeed reachable with a `null` `selectedSite`, additional changes should to be made to protect against runtime errors in the application.

> […] I was able to cancel the purchase just fine despite getting the warning in my console.

I think it's easy to see the line I've reference would error if it is reached with a `null` `selectedSite`. Did you attempt to hit that code path? Are you confident it is unreachable?

> […] Otherwise, my only other idea would be to pass on the fact that we're dealing with a purchase without a site by passing a custom (hardcoded) object that meets the `propType` requirement, but I'm not sure I like that approach.

It may be a technically feasible solution, but I don't like that approach either 🙂 Fighting complexity with complexity is a losing long-term strategy.

Reply:
It looks to me like there are a few places, not just limited to this file, that won't work well if the selected site isn't passed in (not all will be fatals like the above but they will all have consequences).  The ones I found:
- `CancelPurchaseLoadingPlaceholder`
- `CancelPurchase::redirect()`
- `ProductLink`
- `CancelPurchaseForm`

As an example of the effect, you can see that if you go to cancel a purchase with this PR (doesn't have to be a disconnected Jetpack plan; can be a WordPress.com plan or something else) the "View Plan Features" link is missing on the cancellation screen.  That comes from the `ProductLink` component.

It seems like all of the above places don't actually require a site object; they really just require the site slug (and in one case, `CancelPurchaseForm`, the site ID)?  But we already have the site slug, and we can get the site ID from the purchase object (the same way the `ManagePurchase` component already gets it).  So if the components are refactored to expect those properties only, rather than a site object, you could probably make it work that way.  (Not sure about `CancelPurchaseForm` though; that uses the site ID in a tricky way involving an API call via `QuerySitePlans`, and I'm not sure what the consequences are of having that API call be rejected as unauthorized since the user isn't part of the site.)

Reply:
Thank you both for your help with this.

@sirreal asked:
> Did you attempt to hit that code path? Are you confident it is unreachable?

I am somewhat confident this code path is not reachable, as it only appears if the purchase being cancelled is a domain registration, and in these cases we are cancelling a Jetpack registration. I didn't attempt to hit this code path because I'm not sure how I could create an account with a domain registration that is not associated with a site to even reach that code path.

@DavidRothstein suggested:
> It seems like all of the above places don't actually require a site object; they really just require the site slug (and in one case, CancelPurchaseForm, the site ID)? [...] So if the components are refactored to expect those properties only, rather than a site object, you could probably make it work that way.

This makes sense, and it sounds like a good way forward. This way I can keep the propTypes accurate and only require the string that I need, instead of the entire object. On the topic of `QuerySitePlans` I'm a little unsure what this syntax does:

https://github.com/Automattic/wp-calypso/blob/0753a38fc4c61df18e8450abe67b84ad87c7511d/client/components/marketing-survey/cancel-purchase-form/index.jsx#L752

I would interpret that as "If `selectedSite` is present, then include `QuerySitePlans` component". If I'm understanding that correctly, I wouldn't expect this `render` function to generate an error, but it also wouldn't include the component. 

I'd like to include the component, but I'm thinking that would require changing the way we handle disconnected Jetpack purchases by not removing the customer as a user from that site - which might have other side effects. I'll test this but I'm guessing if the correct ID were to be passed to `QuerySitePlans`, the backend would not return a match. If the backend does return a match, perhaps the solution to this issue can be handled entirely on the backend, by returning sites the user is not a member on, but has purchased a Jetpack plan for.